### ISSUE TITLE
Latest Signing Functionality

### ISSUE THREAD

**Issue by robsquireq** [XRPLF/xrpl4j#191]
Hi there, 

I was hoping to receive some support on how exactly I would implement offline signing using the library - Am I taking the right approach in restoring the wallet from the seed and then using the wallet to populate fields in the Payment object builder like below:

![image](https://user-images.githubusercontent.com/80390543/143069669-03727034-ffc7-441f-858c-b2eb1a362c25.png)

I noticed that the functionality I was calling to restore the wallet from the seed and then get the privateKey from it was deprecated, so if you can please point me to how to sign using the private key and also how to do it using just the seed - both of these would be super helpful.

Thanks so much,
Rob

**Comment by nkramer44** [comment]
Hey Rob, what do you mean by "implement offline signing"? When I hear "offline signing", I usually think of storing keys in an external device and signing transactions in that external advice. In that case, you would never have a seed in memory. Since you have a seed in memory here, I think our definitions might be different?

You should be fine to use the `Wallet.privateKey()` field for now, although we are currently in flight with re-writing a lot of this signing functionality. Once we introduce the newest signing functionality, there will be a clear upgrade path on how to get away from this deprecated code. For now though, the way you are using `wallet.privateKey()` is correct and will be supported until version 3.0.0.

**Comment by robsquireq** [comment]
Ah apologies, in this sense i meant offline signing to just be without using an RPC call for example to sign a transaction. The above approach I came across seemed perfect to sign with the library and then later communicate the signed payload to the network.

Thanks for the above clarity...I was able to successfully sign and send a payload using the 'SingleKeySignatureService' with a Private Key which seems to be avoiding the old deprecated code like this:

![image](https://user-images.githubusercontent.com/80390543/143412834-ca0967c3-4afc-4068-9f39-a9fafe9738ac.png)


This seems to work well - I have sent the signed payload to the network and it is successful.

However when I tried to add support for signing using the seed, in order to derive the private key I attempted to use the 'DerivedKeysSignatureService' in order to pass the seed, but it seems in this case the expected Public Key did not match the one generated from the 'DerivedKeysSignatureService'...

![image](https://user-images.githubusercontent.com/80390543/143413742-b122f940-14cd-40b8-9b94-ad5fe0bcc981.png)

I am guessing I have done something a little wrong here - but the approach I took was to generate a wallet using the 'DefaultWalletFactory':

![image](https://user-images.githubusercontent.com/80390543/143414366-4af766a7-ad67-4372-b1e2-6897edf9fa33.png)

And the seed from this I took and put into my derived sign method, where I also specified the 'VerionType' as: ED25519 


I am totally happy to switch to the existing solution to restore the seed to a wallet and get the private key that way, I was just trying to find a long term solution as I was not sure when deprecated code would be getting removed, is 3.0.0 far away?

Thanks again for your help - much appreciated!

**Comment by nkramer44** [comment]
>  in order to derive the private key I attempted to use the 'DerivedKeysSignatureService' in order to pass the seed, but it seems in this case the expected Public Key did not match the one generated from the 'DerivedKeysSignatureService'...

Yes -- The `DerivedKeySignatureService` does not "derive" private keys from an existing XRPL seed. Rather, you can create a `DerivedKeySignatureService` with a random seed value, and this seed value is used to derive an arbitrary number of public/private key pairs, deterministically, given some other value. For example, your system might have existing account UUIDs, which could be passed to the `DerivedKeySignatureService` to deterministically restore a public/private key pair. 

Your initial approach was the correct one. I would guess that v3 is roughly 4-6 months away, so using a `DefaultWalletFactory` to derive a private key, then using that private key in a `SingleKeySignatureService` will be supported at least that long.

**Comment by robsquireq** [comment]
Thanks so much for the response that makes total sense, I will go with the initial approach for the time being! 

Cheers

**Comment by nkramer44** [comment]
No problem! Also, check out the #187 discussion for our proposed roadmap/path to v3. Feel free to comment as well if you have suggestions.
